### ISSUE TITLE
Can not stop, kill, rm, rm --force a container - Docker version 20.10.12, build e91ed57

### ISSUE THREAD

**Issue by D34DC3N73R** [moby/moby#43094]
**Description**

Can not stop, kill, rm, rm --force a container. Still can not stop even after restarting docker.service docker.socket
```
$ docker stop emby
Error response from daemon: cannot stop container: emby: tried to kill container, but did not receive an exit event
```
```
$ docker rm --force emby
Error response from daemon: Could not kill running container e1f8468c9ff5a6a549df76c4e77665c0aa9c2225d94e8ffee331d6514cab956e, cannot remove - tried to kill container, but did not receive an exit event
```

**Steps to reproduce the issue:**
I'm not really sure as I have another system running the same image without any issue

**Describe the results you received:**
Error response from daemon: cannot stop container: emby: tried to kill container, but did not receive an exit event

**Describe the results you expected:**
docker stopping the container

**Additional information you deem important (e.g. issue happens only occasionally):**
When this happens the server also hangs on a reboot command and requires a hard reboot. Upon reboot, the container acts normally for 2-4 hours, but after that, the container behaves as stated in the issue section.

**Output of `docker version`:**

```
$ docker version
Client: Docker Engine - Community
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:45:33 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:42 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.7.1-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)

Server:
 Containers: 26
  Running: 25
  Paused: 0
  Stopped: 1
 Images: 27
 Server Version: 20.10.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.13.0-22-generic
 Operating System: Ubuntu 20.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 20
 Total Memory: 31.13GiB
 Name: admin
 ID: VH3N:4F4H:N36S:L6PF:UBO7:UXK2:EX6V:7LGE:7WSB:7ABI:TAUW:HKI5
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.3 LTS
Release:	20.04
Codename:	focal
```
```
$ uname -a
Linux admin 5.13.0-22-generic #22~20.04.1-Ubuntu SMP Tue Nov 9 15:07:24 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by D34DC3N73R** [comment]
I was able to kill the container with the following process. 

1. Get the container id
`docker ps`
2. Get the containerd-shim process id
`sudo ps awx | grep containerd-shim | grep <<container_id>> | awk '{print $1}'`
3. kill the process id
`sudo kill -9 <<process_id>>

I was then able to `rm --force` the container and recreate it. That did however lead to a corrupted db within the app (I'm not sure if the db was a cause of the docker errors or a result of killing the process). I restored a backup config and haven't run into the issue as of yet - 12+ hours, where I used to reliably get the error in ~ 2 hours. 

Leaving this open until I can test more thoroughly and for anyone else with the same issue. Devs feel free to close if you don't find this relevant.

**Comment by D34DC3N73R** [comment]
Just wanted to follow up and say I'm still seeing the same behavior even after restoring a known working database. Although the time to fail has increased significantly since restoring the DB.

**Comment by D34DC3N73R** [comment]
I rolled back the kernel to 5.11 and the problem has been resolved. I haven't tested any kernels > 5.13, but from the other reports I'm seeing, users with the 5.13 kernel have this issue.

**Comment by johncadengo** [comment]
I am also on 5.13 kernel and wondering if anyone else here has experienced it and if there's a fix?

**Comment by D34DC3N73R** [comment]
This is also an issue with the 5.15 kernel.

**Comment by aojea** [comment]
>  Cgroup Driver: cgroupfs
> Cgroup Version: 1

can be an issue with the cgroup version?

**Comment by ryanmickler** [comment]
I'm also getting this on 5.15 kernel - its a real showstopper.

**Comment by renatoruis** [comment]
`sudo ps awx | grep docker | awk '{print $1}' | xargs kill -9`

**Comment by R8s6** [comment]
I can't believe this is a kernel issue...

tldr: i was on ubuntu 22.04 `linux-generic`, changing to `linux-realtime` fixed it. maybe `linux-generic-hwe` works too.

I came across this thread as i was troubleshooting my sonarr docker on an unbuntu VM, as it has exactly the same symptoms:

1. freeze randomly (always within a day), even though portainer logs show nothing suspecious
2. cannot stop, restart the docker container using cmd
3. "docker compose down" will eventually cause a lingering port that cause "docker compose up" failed, and the lingering port will remain there unless killed.
4. cannot even stop or restart `docker.service`
5. even using `sudo lsof -i -P -n | grep port_number_of_this_docker` will freeze
6. reboot takes like 15 minutes

7. I completely uninstalled docker and removed images:

```
 sudo rm -rf /var/lib/docker
 sudo rm -rf /var/lib/containerd
```
rebooted, and reinstalled docker and everything, boom it was still the same.

It drove my nuts and I was close to reinstall the ubuntu VM.

And thank you guys for pointing out it could be a kernel issue!

I upgraded to latest kernel of the VM and the host, no help. 

Then I changed from `linux-generic` (5.15.0.88.85) to `linux-realtime` (5.15.0-1032), and voila so far so good (a day later). 

I'll observe for a little longer and report back if problem remains (if i don't report back, it means problem solved).

**Comment by wzin** [comment]
Observing the same problem:
- kernel `6.2.0-37-generic`
- docker version `Docker version 24.0.7, build afdd53b`
- logs:

```bash
Dec  7 15:35:37 foobar2-vm1 dockerd[2062066]: time="2023-12-07T15:35:37.192457525Z" level=info msg="Container failed to exit within 10s of signal 15 - using the force" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f
Dec  7 15:36:09 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:36:09.682689689Z" level=info msg="Container failed to exit within 10s of signal 15 - using the force" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f
Dec  7 15:36:19 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:36:19.693315815Z" level=error msg="Container failed to exit within 10s of kill - trying direct SIGKILL" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="context deadline exceeded"
Dec  7 15:36:23 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:36:23.696084091Z" level=error msg="error killing container: context deadline exceeded" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="tried to kill container, but did not receive an exit event"
Dec  7 15:36:23 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:36:23.696190451Z" level=error msg="error shutting down container" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="Failed to stop container 15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f with error: tried to kill container, but did not receive an exit event"
Dec  7 15:38:42 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:38:42.462977310Z" level=info msg="Container failed to exit within 10s of signal 15 - using the force" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f
Dec  7 15:38:52 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:38:52.467377343Z" level=error msg="Container failed to exit within 10s of kill - trying direct SIGKILL" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="context deadline exceeded"
Dec  7 15:38:56 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:38:56.468548754Z" level=error msg="error killing container: context deadline exceeded" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="tried to kill container, but did not receive an exit event"
Dec  7 15:38:56 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:38:56.468796223Z" level=error msg="Handler for POST /v1.43/containers/15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f/stop returned error: cannot stop container: 15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f: tried to kill container, but did not receive an exit event"
Dec  7 15:42:12 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:42:12.390963194Z" level=info msg="Container failed to exit within 10s of signal 15 - using the force" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f
Dec  7 15:42:22 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:42:22.395436889Z" level=error msg="Container failed to exit within 10s of kill - trying direct SIGKILL" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="context deadline exceeded"
Dec  7 15:42:26 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:42:26.397120568Z" level=error msg="error killing container: context deadline exceeded" container=15cbe4a91c773d636682e67952b163056df17a4a93cac74c74916ae2c2eb2e0f error="tried to kill container, but did not receive an exit event"
Dec  7 15:42:26 foobar2-vm1 dockerd[381308]: time="2023-12-07T15:42:26.397322801Z" level=error msg="Handler for POST /v1.43/containers/15cbe4a91c77/stop returned error: cannot stop container: 15cbe4a91c77: tried to kill container, but did not receive an exit event"
```

All this happened when my vm ran out of memory so this is not an expected sittuation that I ran into, but docker should let me kill the container nevertheless. Also - I stopped the demon and restarted it - it did not help because the "hanging" containers re-appeared  after restart.

**Comment by iec61131** [comment]
same system on my side:
kernel 6.2.0-37-generic
docker version Docker version 24.0.7, build afdd53b

Any type of docker container cannot be stopped gracefully. Any docker bug tracking tool we can report upstream?

**Comment by tonytaylor85** [comment]
I started experiencing this problem after I updated my proxmox host. I'm running in LXC with nvidia and coral passthru.
from-
kernel 6.5.13-3-pve
docker 23.0.5-bc4487a

to-
kernel 6.5.13-5-pve
docker 26.0.1-d260a54

**Comment by zhangguanzhang** [comment]
same issue with centos7.9 kernel 3.10

**Comment by JohnGalt1717** [comment]
Still having this with fully patched Ubunto Server 22.04

**Comment by mikeygithub** [comment]
same issue with  Kernel Version: 6.5.0-18-generic, Ubuntu: 22.04.4 LTS, Docker Engine: 26.1.0

**Comment by cheriL** [comment]
same issue with Kernel Version: 5.15.0-124-generic, Ubuntu 22.04.5 LTS, Docker Engine: 27.3.1

**Comment by serhii-chernenko** [comment]
Same issue from 5.15.0-86-generic

**Comment by ShivamB25** [comment]
wrote a script that does the killing. feel free to use

```bash
#!/bin/bash

# Script to stop Docker containers gracefully or forcefully
# Usage: ./kill-container.sh <container_name>

# Text colors
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
NC='\033[0m' # No Color

# Function to print messages
print_info() {
    echo -e "${YELLOW}[INFO]${NC} $1"
}

print_success() {
    echo -e "${GREEN}[SUCCESS]${NC} $1"
}

print_error() {
    echo -e "${RED}[ERROR]${NC} $1"
}

# Check if container name is provided
if [ $# -eq 0 ]; then
    print_error "No container name provided"
    echo "Usage: $0 <container_name>"
    exit 1
fi

CONTAINER_NAME=$1

# Check if container exists
if ! docker ps | grep -q $CONTAINER_NAME; then
    print_error "Container '$CONTAINER_NAME' not found in running containers"
    echo "Running containers:"
    docker ps --format "table {{.Names}}\t{{.Status}}"
    exit 1
fi

print_info "Attempting to stop container '$CONTAINER_NAME'..."

# First try: docker stop
if docker stop $CONTAINER_NAME; then
    print_success "Successfully stopped container '$CONTAINER_NAME'"
    exit 0
fi

# Second try: docker kill
print_info "Docker stop failed, attempting docker kill..."
if docker kill $CONTAINER_NAME; then
    print_success "Successfully killed container '$CONTAINER_NAME'"
    exit 0
fi

# Last resort: find and kill containerd-shim process
print_info "Docker kill failed, attempting to kill containerd-shim process..."

# Get container ID
CONTAINER_ID=$(docker ps | grep $CONTAINER_NAME | awk '{print $1}')

if [ -z "$CONTAINER_ID" ]; then
    print_error "Could not find container ID for '$CONTAINER_NAME'"
    exit 1
fi

# Get containerd-shim process ID
SHIM_PID=$(sudo ps awx | grep containerd-shim | grep $CONTAINER_ID | awk '{print $1}')

if [ -z "$SHIM_PID" ]; then
    print_error "Could not find containerd-shim process"
    exit 1
fi

# Kill the process
print_info "Killing process $SHIM_PID..."
sudo kill -9 $SHIM_PID

print_info "Process killed. Checking container status..."
sleep 2 # Give it a moment to update

# Final status check
if ! docker ps | grep -q $CONTAINER_NAME; then
    print_success "Container '$CONTAINER_NAME' successfully stopped"
else
    print_error "Container '$CONTAINER_NAME' is still running"
    docker ps | grep $CONTAINER_NAME
fi

exit 0```
